File: NX2312/UGOPEN/SampleNXOpenApplications/Java/Routing/Routing_Example_Spool_Plugins.java
/*==================================================================================================

    This material contains trade secrets or otherwise confidential information owned by
    Siemens Industry Software Inc. or its affiliates (collectively, "SISW"), or its licensors.
    Access to and use of this information is strictly limited as set forth in the Customer's
    applicable agreements with SISW.

    Unpublished work Copyright 2023 Siemens

====================================================================================================

    NOTE:  NX Development provides programming examples for illustration only.
    NX Development assumes you are familiar with the programming language
    demonstrated in these examples, and the tools used to create and debug NX/Open
    programs. GTAC support professionals can help explain the functionality of
    a particular procedure, but neither GTAC nor NX Development will modify
    these examples to provide added functionality or construction procedures.

File description:

    A Java class that implements example Routing Spool plugins.


==================================================================================================*/
import java.rmi.RemoteException;

import nxopen.NXException;
import nxopen.NXObject;
import nxopen.Part;
import nxopen.Session;
import nxopen.SessionFactory;
import nxopen.assemblies.Component;
import nxopen.mechanicalrouting.*;
import nxopen.routingcommon.CustomManager.SpoolNamePlugin;
import nxopen.schematic.mechanical.Run;

public class Routing_Example_Spool_Plugins
    implements SpoolNamePlugin
{
    static Routing_Example_Spool_Plugins thePlugins    = null;
    static Session                       session       = null;

    //----------------------------------------------------------------------------------------------
    // Called when NX starts.
    // Creates a new instance of this Callbacks class.
    public static void startup( String [] args ) throws NXException, java.rmi.RemoteException
    {
        if (session == null)
            session = (Session)SessionFactory.get("Session");
    
        // Calling CreateRoutingSession is required before registering Routed System Designer (RSD)
        // plugins, callbacks, or design rules to ensure the Routing libraries are properly loaded.
        session.createRoutingSession();
    
        thePlugins = new Routing_Example_Spool_Plugins();
    
        session.licenseManager().releaseAll( null );
    }
    
    //----------------------------------------------------------------------------------------------
    // The constructor for this class.
    // Registers the Spool Plugins.
    public Routing_Example_Spool_Plugins() throws RemoteException, NXException
    {
        nxopen.routingcommon.CustomManager customManager = session.routingCustomManager();
    
        customManager.setSpoolNamePlugin( this );
    }
    
    //----------------------------------------------------------------------------------------------
    // Called by the Spool dialog to create a name for the new spool.
    // 
    // The name this plugin generates is a combination of:
    // 
    //   - Prefix:        The spool Name Prefix defined in the Routed System Designer->Mechanical->Spools customer defaults.
    //   - Block:         The Block identifier is simply the name of the work part.
    //   - Specification: The specification assigned to the components.
    //   - Run ID:        The identifier of the run to which the components belong.
    //   - Index:         An index one greater than the number of existing spools in the part.
    //
    // \param[in]
    //      spoolPart
    //          The part in which the new spool is created.
    //
    // \param[in]
    //      spoolComponents
    //          The connected components that will make up the new spool.
    //
    // \return
    //      The name the Routing application will use for the new spool.
    public String spoolNamePlugin
    (
        Part         spoolPart,
        Component[] spoolComponents
    ) throws NXException, RemoteException
    {
        String prefix    = session.optionsManager().getStringValue( "RSD_SpoolNamePrefix" );
        String blockName = getBlock( spoolPart );
        
        String strings[]     = getSpecificationAndRun( spoolComponents );
        String specification = strings[0];
        String runId           = strings[1];
    
        int nSpools = 0;
        SpoolCollection.Iterator spoolsIterator = spoolPart.mechanicalRoutingCollectionsManager().spools().iterator();
        while (spoolsIterator.hasNext())
        {
            ++nSpools;
            spoolsIterator.next();
        }
        
        String generatedName = "";
        if ( prefix != null && !prefix.isEmpty() )
            generatedName = prefix;
    
        if ( blockName != null && !blockName.isEmpty() )
            generatedName += blockName + " ";
    
        if ( specification != null && !specification.isEmpty() )
            generatedName += specification + " ";
    
        if ( runId != null && !runId.isEmpty() )
            generatedName += runId + " ";
    
        generatedName += String.format("%02d", nSpools);
        
        return generatedName;
    }
    
    //----------------------------------------------------------------------------------------------
    // Returns the spool part's name as the "Block" to add to the spool name.
    private String getBlock
    (
        Part spoolPart
    ) throws RemoteException, NXException
    {
        return spoolPart.name();
    }
    
    //----------------------------------------------------------------------------------------------
    // Gets the specification and the run from the components selected for the spool.
    private String[] getSpecificationAndRun
    (
        Component[] spoolComponents
    ) throws RemoteException, NXException
    {
        String specification = "";
        String runId = "";
        
        RunsAssignmentManager runsManager = RoutingManagerFactory.GetRoutingManager( session ).runsAssignmentManager();
    
        for ( Component spoolComponent : spoolComponents )
        {
            String componentSpecification = getStringAttributeValueFromObject( spoolComponent, "SPECIFICATION" );
            if ( componentSpecification != null && !componentSpecification.isEmpty() )
                specification = componentSpecification;
            
            String componentRun = getRun( runsManager, spoolComponent );
            if (componentRun != null && !componentRun.isEmpty() )
                runId = componentRun;
            
            if (specification != null && !specification.isEmpty() &&
                runId         != null && !runId.isEmpty() )
            {
                break;
            }
        }
        
        String[] strings = new String[2];
        strings[0] = specification;
        strings[1] = runId;
    
        return strings;
    }
    
    //----------------------------------------------------------------------------------------------
    // Uses the runs assignment manager to figure out which run this component belongs to.
    private static String getRun
    (
        RunsAssignmentManager runsManager,
        Component             component
    ) throws RemoteException, NXException
    {
        if (runsManager == null)
            return "";
    
        Run[] runs = runsManager.getAssignedRuns( component );
        for( Run run : runs )
        {
            String runId = run.identifier();
            if (runId != null && !runId.isEmpty() )
                return runId;
        }
    
        return "";
    }
    
    //----------------------------------------------------------------------------------------------
    // Returns the string value of the attribute or null if it does not exist.
    private static String getStringAttributeValueFromObject
    (
        NXObject nxObject,
        String   attributeName
    ) throws RemoteException, NXException
    {
        NXObject.AttributeInformation attributeInformation = getAttributeFromObject( nxObject, attributeName );
        if ( attributeInformation == null || attributeInformation.type != NXObject.AttributeType.STRING )
            return null;
    
        return attributeInformation.stringValue;
    }
    
    //----------------------------------------------------------------------------------------------
    // Returns true and the attribute information object of the attribute if the attribute exists.
    // Returns false if not.
    private static NXObject.AttributeInformation getAttributeFromObject
    (
        NXObject nxObject,
        String   attributeName
    ) throws RemoteException, NXException
    {
        return getAttributeFromList( nxObject.getUserAttributes(), attributeName );
    }
    
    //----------------------------------------------------------------------------------------------
    // Returns true and the attribute information object of the attribute if the attribute is in the list.
    // Returns false if not.
    // NOTE: The attribute name comparison is case insensitive.
    private static NXObject.AttributeInformation getAttributeFromList
    (
        NXObject.AttributeInformation[] attributes,
        String                          attributeName
    )
    {
        for ( nxopen.NXObject.AttributeInformation attribute : attributes )
        {
            // Watch out for attributes with no title.
            if ( attribute.title.length() == 0 )
                continue;
    
            // NOTE: This does not look for aliases on the attribute titles.
            if ( attribute.title.compareToIgnoreCase( attributeName ) == 0 )
            {
                return attribute;
            }
        }
    
        return null;
    }
    
    //----------------------------------------------------------------------------------------------
    // Tells NX when to unload your application.
    // For Routing callbacks, plugins, and Design Rules, this MUST return AtTermination.
    public static int getUnloadOption(String arg)
    {
        return Session.LibraryUnloadOption.AT_TERMINATION;
    }
}

File: NX2412/UGOPEN/SampleNXOpenApplications/DotNet/Routing/Routing_Example_Spool_Plugins.cs
/*==================================================================================================

    This material contains trade secrets or otherwise confidential information owned by
    Siemens Industry Software Inc. or its affiliates (collectively, "SISW"), or its licensors.
    Access to and use of this information is strictly limited as set forth in the Customer's
    applicable agreements with SISW.

    Unpublished work Copyright 2023 Siemens

====================================================================================================

    NOTE:  NX Development provides programming examples for illustration only.
    NX Development assumes you are familiar with the programming language
    demonstrated in these examples, and the tools used to create and debug NX/Open
    programs. GTAC support professionals can help explain the functionality of
    a particular procedure, but neither GTAC nor NX Development will modify
    these examples to provide added functionality or construction procedures.

File description:

    A C# class that implements example Routing Spool plugins.


==================================================================================================*/
using System;
using System.IO;
using NXOpen;
using NXOpen.Assemblies;
using NXOpen.MechanicalRouting;
using NXOpen.RoutingCommon;
using NXOpen.Schematic.Mechanical;
using NXOpen.UF;
using NXOpen.PDM;

namespace RoutingCommon
{
    public class SpoolPlugins
    {
        // Static class members
        static Session       session       = null;
        static ListingWindow listingWindow = null;
        static CustomManager customManager = null;
        private static int spoolAssemblyCounter = 0;

        //------------------------------------------------------------------------------------------
        // Called when NX starts up to register the plugins.
        public static int Startup()
        {
            if (session == null)
                session = Session.GetSession();

            if (listingWindow == null)
                listingWindow = session.ListingWindow;

            // Calling CreateRoutingSession is required before registering Routed System Designer (RSD)
            // plugins, callbacks, or design rules to ensure the Routing libraries are properly loaded.
            session.CreateRoutingSession();

            if (customManager == null)
                customManager = session.RoutingCustomManager;

            customManager.SetSpoolNamePlugin( GenerateSpoolName );
            customManager.SetSpoolAssemblyNamePlugin( GenerateSpoolAssemblyName );

            session.LicenseManager.ReleaseAll( null );

            return 0;
        }

        //------------------------------------------------------------------------------------------
        // Called by the Spool dialog to create a name for the new spool.
        // 
        // The name this plugin generates is a combination of:
        // 
        //   - Spool description: See BuildSpoolDescription for more information.
        //   - Index:             An index one greater than the number of existing spools in the part.
        //
        // \param[in]
        //      spoolPart
        //          The part in which the new spool is created.
        //
        // \param[in]
        //      spoolComponents
        //          The connected components that will make up the new spool.
        //
        // \return
        //      The name the Routing application will use for the new spool.
        public static string GenerateSpoolName
        (
            NXOpen.Part spoolPart,
            Component[] spoolComponents
        )
        {
            string generatedName = BuildSpoolDescription( spoolPart, spoolComponents );

            Spool[] spools = spoolPart.MechanicalRoutingCollectionsManager.Spools.ToArray();
            generatedName += spools.Length.ToString( "D2" );

            string message = generatedName;
            message += " was created.";
            listingWindow.Open();
            listingWindow.WriteFullline( message );

            return generatedName;
        }

        //------------------------------------------------------------------------------------------
        // Called by the Publish Spool Assemblies command to create a name for the new spool assembly part.
        //
        // The spool assembly name plugin is given a data object that needs to be filled out
        // completely, either for a native file or a Teamcenter item, in order for Routing to
        // create a new part for the spool.
        //
        // \param[in]
        //      spoolComponents
        //          The components of the spool that are to go in the new spool assembly.
        // 
        // \param[out]
        //      dataObject
        //          The data object that is to be filled out by the plugin.
        public static void GenerateSpoolAssemblyName
        (
            Component[]                           spoolComponents,
            NXOpen.Routing.AssemblyNamePluginData dataObject
        )
        {
            ++spoolAssemblyCounter;

            string assemblyName = "SpoolAssembly_" + spoolAssemblyCounter.ToString( "D4" );

            if (session.PdmSession == null )
            {
                dataObject.NativeFileName = assemblyName;
                dataObject.NativePath     = ""; // An empty path means the current directory.
                return;
            }

            UFUgmgr ugmgr = UFSession.GetUFSession().Ugmgr;

            Tag rootFolderTag;
            ugmgr.AskRootFolder( out rootFolderTag );

            // Find the home folder and then append the name of the folder that should hold the spool assembly.
            string homeFolder = "";
            ugmgr.AskFolderName( rootFolderTag, out homeFolder );
            dataObject.ManagedModeFolderName      = homeFolder + ":SpoolAssemblies";

            dataObject.ManagedModeItemName        = ""; // Return an empty item name to get the next available item ID.
            dataObject.ManagedModeItemRevision    = "A";
            dataObject.ManagedModeItemType        = "Item";
            dataObject.ManagedModeItemDescription = GetSpoolName( spoolComponents );
        }

        //------------------------------------------------------------------------------------------
        // Returns the name of the spool that owns the given components.
        //
        // This assumes all components belong to the same spool and only belong to one spool.
        //
        private static string GetSpoolName
        (
            Component[] spoolComponents
        )
        {
            if (spoolComponents.Length == 0)
                return "";

            Session         session         = Session.GetSession();
            Part            workPart        = session.Parts.Work;
            SpoolCollection spoolCollection = workPart.MechanicalRoutingCollectionsManager.Spools;

            Spool spool = spoolCollection.GetSpoolFromComponent( spoolComponents[0] );
            if ( spool == null )
                return "";

            return spool.Name;
        }

        //------------------------------------------------------------------------------------------
        // Builds a description of the spool that includes:
        // 
        //   - Prefix:        The spool Name Prefix defined in the Routed System Designer->Mechanical->Spools customer defaults.
        //   - Block:         The Block identifier is simply the name of the work part.
        //   - Specification: The specification assigned to the components.
        //   - Run ID:        The identifier of the run to which the components belong.
        //
        private static string BuildSpoolDescription
        (
            NXOpen.Part spoolPart,
            Component[] spoolComponents
        )
        {
            string prefix = session.OptionsManager.GetStringValue( "RSD_SpoolNamePrefix" );
            string blockName = GetBlock( spoolPart );

            string specification;
            string runId;
            GetSpecificationAndRun( spoolComponents, out specification, out runId );

            string description = "";
            if (!String.IsNullOrEmpty( prefix ))
                description = prefix;

            if (!String.IsNullOrEmpty( blockName ))
                description += blockName + " ";

            if (!String.IsNullOrEmpty( specification ))
                description += specification + " ";

            if (!String.IsNullOrEmpty( runId ))
                description += runId + " ";

            return description;
        }

        //------------------------------------------------------------------------------------------
        // Returns the spool part's name as the "Block" to add to the spool name.
        private static string GetBlock
        (
            NXOpen.Part spoolPart
        )
        {
            if ( spoolPart != null )
                return Path.GetFileName( spoolPart.Name );

            return "";
        }

        //------------------------------------------------------------------------------------------
        // Gets the specification and the run from the components selected for the spool.
        private static void GetSpecificationAndRun
        (
            Component[] spoolComponents,
            out string  specification,
            out string  runId
        )
        {
            specification = "";
            runId         = "";

            RunsAssignmentManager runsManager = NXOpen.MechanicalRouting.RoutingManager.GetRoutingManager( session ).RunsAssignmentManager;

            foreach ( Component spoolComponent in spoolComponents )
            {
                string componentSpecification = GetStringAttributeValueFromObject( spoolComponent, "SPECIFICATION" );
                if (!String.IsNullOrEmpty( componentSpecification ))
                    specification = componentSpecification;

                string componentRun = GetRun( runsManager, spoolComponent );
                if (!String.IsNullOrEmpty( componentRun ))
                    runId = componentRun;

                if (!String.IsNullOrEmpty( componentSpecification ) && !String.IsNullOrEmpty( componentRun ))
                    break;
            }
        }

        //--------------------------------------------------------------------------------------------------
        // Uses the runs assignment manager to figure out which run this component belongs to.
        private static string GetRun
        (
            RunsAssignmentManager runsManager,
            Component             component
        )
        {
            if (runsManager == null)
                return "";

            Run[] runs = runsManager.GetAssignedRuns( component );
            foreach ( Run run in runs )
            {
                string runId = run.Identifier;
                if (!String.IsNullOrEmpty( runId ))
                    return runId;
            }

            return "";
        }

        //------------------------------------------------------------------------------------------
        private static String GetStringAttributeValueFromObject
        (
            NXObject nxObject,
            string   attributeName
        )
        {
            NXObject.AttributeInformation attributeInformation;

            if (GetAttributeFromObject( nxObject, attributeName, out attributeInformation ))
                return attributeInformation.StringValue;

            return null;
        }

        //------------------------------------------------------------------------------------------
        private static bool GetAttributeFromObject
        (
            NXObject                          nxObject,
            string                            attributeName,
            out NXObject.AttributeInformation attributeInformation
        )
        {
            return GetAttributeFromList( nxObject.GetUserAttributes(), attributeName, out attributeInformation );
        }

        //------------------------------------------------------------------------------------------
        private static bool GetAttributeFromList
        (
            NXObject.AttributeInformation[]   attributes,
            string                            attributeName,
            out NXObject.AttributeInformation attributeInformation
        )
        {
            attributeInformation = new NXObject.AttributeInformation();

            foreach (NXObject.AttributeInformation attribute in attributes)
            {
                string titleString = attribute.Title;

                // Watch out for attributes with no title.
                if (String.IsNullOrEmpty( titleString ))
                    continue;

                // NOTE: This does not look for aliases on the attribute titles.
                if (titleString == attributeName)
                {
                    attributeInformation = attribute;
                    return true;
                }
            }

            return false;
        }

        //------------------------------------------------------------------------------------------
        // Tells NX when to unload your application.
        // For Routing callbacks, plugins, and Design Rules, this MUST return AtTermination.
        public static int GetUnloadOption( string arg )
        {
            return System.Convert.ToInt32( Session.LibraryUnloadOption.AtTermination );
        }
    }
}
